lib/migrations/migrations.js

```diff
@@ -0,0 +1,350 @@
+
+      await this.saveState(state);
+      return;
+    }
+
+    const lastID = this.getLastMigrationID();
+    const state = await this.getState();
+
+    this.logger.debug('Last migration %d, last available migration: %d',
+      state.lastMigration, lastID);
+
+    for (const id of state.skipped) {
+      const skippedMigration = new this.migrations[id](this.options);
+      this.logger.warning(skippedMigration.warning());
+    }
+
+    if (state.inProgress) {
+      this.logger.info('Continue progress on migration: ',
+        state.lastMigration + 1);
+    }
+
+    let hasMigrated = false;
+    for (let i = state.lastMigration + 1; i <= lastID; i++) {
+      const currentMigration = new this.migrations[i](this.options);
+      const type = await currentMigration.check();
+
+      switch (type)  {
+        case types.FAKE_MIGRATE: {
+          this.logger.info('Does not apply, fake migrating %d.', i);
+          state.lastMigration = i;
+          await this.saveState(state);
+          break;
+        }
+        case types.SKIP: {
+          this.logger.info('Can not migrate, skipping %d.', i);
+          this.logger.warning(currentMigration.warning());
+          state.lastMigration = i;
+          state.skipped.push(i);
+          await this.saveState(state);
+          break;
+        }
+        case types.MIGRATE: {
+          if (!this.options.migrateFlag)
+            throw new Error('Database needs migration.');
+
+          state.inProgress = true;
+          await this.saveState(state);
+
+          this.logger.info('Migration %d in progress..', i);
+          const batch = this.ldb.batch();
+          await currentMigration.migrate(batch);
+
+          state.inProgress = false;
+          state.lastMigration = i;
+          this.writeState(batch, state);
+
+          await batch.write();
+          this.logger.info('Migration %d is done.', i);
+          hasMigrated = true;
+          break;
+        }
+        default:
+          throw new Error('Unknown migration type.');
+      }
+    }
+
+    if (!hasMigrated && this.options.migrateFlag)
+      throw new Error('There are no available migrations.');
+  }
+
+  /**
+   * Ensure we have migration entry in DB.
+   * Do migration of the old migration states.
+   */
+
+  async ensure() {
+    const migrations = await this.ldb.get(this.layout.M.encode());
+
+    // This could mean two things:
+    //  1. database is new.
+    //  2. we are migrating from old version.
+    if (!migrations) {
+      const b = this.ldb.batch();
+      const state = new MigrationState();
+
+      // move migration entries if we are migrating
+      // from old version with Migrations.
+      // p.s. if there are no migrations:
+      //  1. database is new
+      //  2. we are migrating from even older version.
+      const oldMigrations = await this.ldb.keys({
+        gte: oldLayout.M.min(),
+        lte: oldLayout.M.max(),
+        parse: key => oldLayout.M.decode(key)[0]
+      });
```

Comment:
I find it a little funny that this new Migrations module... requires a migration. Are there any failure states where this migration of `layout.M` gets interrupted or incomplete? I wonder if the new migrations DB prefix should be different so it's always fresh for this upgrade, and we can just mark `layout.M` as `// deprecated`.

Reply:
We can definitely mark layout.M deprecated and use different prefix for this.
We currently have single migration in both wallet and chain dbs, so instead of iteration I can just check if layout.M exists - then we have done THAT ONE migration so skip it. It can make code somewhat cleaner.

I just did not want to leave unused prefix in the layout and thought this was not that much complicated, but it probably is